src/bitcoin-cli.cpp

```diff
@@ -274,19 +274,49 @@ struct BaseRequestHandler {
 
 /** Process addrinfo requests */
 struct AddrinfoRequestHandler : BaseRequestHandler {
+    static constexpr int ID_FULL_ADDRMAN = 0;
+    static constexpr int ID_FILTERED_ADDRMAN = 1;
+
+    /** Prepare a batch of JSON-RPC requests to retrieve addrman data.
+     *  This combines:
+     *   - getaddrmaninfo (available from v26.0+) for stats of all addresses in the addrman (used for selecting peers)
+     *   - getnodeaddresses (v22.0+) for stats of filtered address in the addrman (used by old -addrinfo and kept for not breaking user space)
+     */
     UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& args) override
     {
         if (!args.empty()) {
             throw std::runtime_error("-addrinfo takes no arguments");
         }
+
+        UniValue result(UniValue::VARR);
+        result.push_back(JSONRPCRequestObj("getaddrmaninfo", NullUniValue, ID_FULL_ADDRMAN));
         UniValue params{RPCConvertValues("getnodeaddresses", std::vector<std::string>{{"0"}})};
-        return JSONRPCRequestObj("getnodeaddresses", params, 1);
+        result.push_back(JSONRPCRequestObj("getnodeaddresses", params, ID_FILTERED_ADDRMAN));
+        return result;
     }
 
-    UniValue ProcessReply(const UniValue& reply) override
+    UniValue ProcessReply(const UniValue &batch_in) override
     {
-        if (!reply["error"].isNull()) return reply;
-        const std::vector<UniValue>& nodes{reply["result"].getValues()};
+        const std::vector<UniValue> batch = JSONRPCProcessBatchReply(batch_in);
+        UniValue result{UniValue::VOBJ};
+
+        // Process getaddrmaninfo reply (only if RPC exists)
+        if (batch[ID_FULL_ADDRMAN]["error"].isNull()) {
+            UniValue addresses{UniValue::VOBJ};
+            const std::vector<std::string>& network_types{batch[ID_FULL_ADDRMAN]["result"].getKeys()};
+            const std::vector<UniValue>& addrman_counts{batch[ID_FULL_ADDRMAN]["result"].getValues()};
+            for (size_t i = 0; i < network_types.size() - 1; ++i) {
+                uint64_t addr_count = addrman_counts[i]["total"].getInt<int>();
+                addresses.pushKV(network_types[i], addr_count);
+            }
+            uint64_t total = addrman_counts[network_types.size() - 1]["total"].getInt<int>();
+            addresses.pushKV("total", total);
```

Comment:
This assumes that the `all_networks` key is the last key in the JSON returned by `getaddrmaninfo` but JSON objects don't have order (unlike arrays). So it is possible that the order changes any time (in theory) and then this code will break. Better write it in such a way that it works regardless of the order in which keys are returned.

Maybe something like:
```
iterate all from i = 0 to i < network_types.size()
    if key is all_networks
        this is the total
    else
        this is one of the networks